scripts/publish-plugin.sh

```diff
@@ -24,20 +24,46 @@ fi
 # Tar up the plugin
 tar -czf ${PLUGIN_PACKAGE_PATH} -C ${WORK_PATH} .
 
+# Push the current AWS credentials, since we'll need them to assume a second role.
+export INITIAL_AWS_ACCESS_KEY_ID="${AWS_ACCESS_KEY_ID}"
+export INITIAL_AWS_SECRET_ACCESS_KEY="${AWS_SECRET_ACCESS_KEY}"
+
 # rel.pulumi.com is in our production account, so assume that role first
 CREDS_JSON=$(aws sts assume-role \
                  --role-arn "arn:aws:iam::058607598222:role/UploadPulumiReleases" \
                  --role-session-name "upload-plugin-pulumi-resource-vsphere" \
                  --external-id "upload-pulumi-release")
 
-# Use the credentials we just assumed
+# Use the credentials we just assumed.
 export AWS_ACCESS_KEY_ID=$(echo ${CREDS_JSON}     | jq ".Credentials.AccessKeyId" --raw-output)
 export AWS_SECRET_ACCESS_KEY=$(echo ${CREDS_JSON} | jq ".Credentials.SecretAccessKey" --raw-output)
 export AWS_SECURITY_TOKEN=$(echo ${CREDS_JSON}    | jq ".Credentials.SessionToken" --raw-output)
 
-echo "Uploading ${PLUGIN_PACKAGE_NAME}..."
+echo "Uploading ${PLUGIN_PACKAGE_NAME} to s3://rel.pulumi.com..."
 
 aws s3 cp --only-show-errors "${PLUGIN_PACKAGE_PATH}" "s3://rel.pulumi.com/releases/plugins/${PLUGIN_PACKAGE_NAME}"
 
+# Assume the role to publish plugins to s3://get.pulumi.com. We upload the plugins to two buckets while
+# we transition to only publishing/serving them from get.pulumi.com.
+echo "Uploading ${PLUGIN_PACKAGE_NAME} to s3://get.pulumi.com..."
+
+# Restore the initial AWS credentials we had, since the assumed role doesn't have the
+# ability to assume this other role to publish into a different bucket.
+unset {AWS_ACCESS_KEY_ID,AWS_SECRET_ACCESS_KEY,AWS_SECURITY_TOKEN}
+export AWS_ACCESS_KEY_ID="${INITIAL_AWS_ACCESS_KEY_ID}"
+export AWS_SECRET_ACCESS_KEY="${INITIAL_AWS_SECRET_ACCESS_KEY}"
+
+CREDS_JSON=$(aws sts assume-role \
+                 --role-arn "arn:aws:iam::058607598222:role/PulumiUploadRelease" \
+                 --role-session-name "upload-plugin-pulumi-resource-vsphere" \
+                 --external-id "upload-pulumi-release")
+export AWS_ACCESS_KEY_ID=$(echo ${CREDS_JSON}     | jq ".Credentials.AccessKeyId" --raw-output)
+export AWS_SECRET_ACCESS_KEY=$(echo ${CREDS_JSON} | jq ".Credentials.SecretAccessKey" --raw-output)
+export AWS_SECURITY_TOKEN=$(echo ${CREDS_JSON}    | jq ".Credentials.SessionToken" --raw-output)
+
+aws s3 cp \
+    --only-show-errors --acl public-read \
+    "${PLUGIN_PACKAGE_PATH}" "s3://get.pulumi.com/releases/plugins/${PLUGIN_PACKAGE_NAME}"
+
```

Comment:
Don't we need to reset the envvars back to `INITIAL_AWS_ACCESS_KEY_ID` and `INITIAL_AWS_SECRET_ACCESS_KEY` after pushing to `get.pulumi.com` above?

Reply:
Strictly speaking, no. My understanding is that unless the script is "sourced", any environment variables it changes won't be persisted as soon as the process ends. (So no mater what state `publish-plugin.sh` leaves the AWS env vars in, it won't impact the caller script unless it ran `. publish-plugin.sh`.)

But it is arguably sloppy not to restore them. Happy to make that change, the only reason not to is to just keep the script more concise. (But it is a probably waiting to happen.)

Reply:
I'll just go ahead and make the change here, and update the script I'm using to generate the diff. (So that we don't replicate the same code smell across all providers...)